examples/launcher-templates/helm/2.3.0-v1/job.tpl

```diff
@@ -0,0 +1,272 @@
+# Version: 2.3.0
+# DO NOT MODIFY the "Version: " key
+# Helm Version: v1
+{{- $templateData := include "rstudio-library.templates.data" nil | mustFromJson }}
+apiVersion: batch/v1
+kind: Job
+metadata:
+  annotations:
+    {{- with .Job.metadata.job.annotations }}
+    {{- range $key, $val := . }}
+    {{ $key }}: {{ toYaml $val | indent 4 | trimPrefix (repeat 4 " ") }}
+    {{- end }}
+    {{- end }}
+    {{- with $templateData.job.annotations }}
+    {{- range $key, $val := . }}
+    {{ $key }}: {{ toYaml $val | indent 4 | trimPrefix (repeat 4 " ") }}
+    {{- end }}
+    {{- end }}
+  labels:
+    app.kubernetes.io/managed-by: "launcher"
+    {{- with .Job.metadata.job.labels }}
+    {{- range $key, $val := . }}
+    {{ $key }}: {{ toYaml $val | indent 4 | trimPrefix (repeat 4 " ") }}
+    {{- end }}
+    {{- end }}
+    {{- with $templateData.job.labels }}
+    {{- range $key, $val := . }}
+    {{ $key }}: {{ toYaml $val | indent 4 | trimPrefix (repeat 4 " ") }}
+    {{- end }}
+    {{- end }}
+  generateName: {{ toYaml .Job.generateName }}
+spec:
+  backoffLimit: 0
+  template:
+    metadata:
+      annotations:
+        {{- if .Job.tags }}
+        {{- $i := 0 }}
+        {{- range .Job.tags }}
+        USER_TAG_{{ $i }}: {{ toYaml . | indent 8 | trimPrefix (repeat 8 " ") }}
+        {{- $i = add $i 1 }}
+        {{- end }}
+        {{- end }}
+        stdin: {{ toYaml .Job.stdin | indent 8 | trimPrefix (repeat 8 " ") }}
+        user: {{ toYaml .Job.user }}
+        name: {{ toYaml .Job.name }}
+        service_ports: {{ toYaml .Job.servicePortsJson }}
+        {{- if .Job.metadata }}
+        user_metadata: {{ toJson .Job.metadata | toYaml | indent 8 | trimPrefix (repeat 8 " ") }}
+        {{- end }}
+        {{- with .Job.metadata.pod.annotations }}
+        {{- range $key, $val := . }}
+        {{ $key }}: {{ toYaml $val | indent 8 | trimPrefix (repeat 8 " ") }}
+        {{- end }}
+        {{- end }}
+        {{- with $templateData.pod.annotations }}
+        {{- range $key, $val := . }}
+        {{ $key }}: {{ toYaml $val | indent 8 | trimPrefix (repeat 8 " ") }}
+        {{- end }}
+        {{- end }}
+      labels:
+        {{- with .Job.metadata.pod.labels }}
+        {{- range $key, $val := . }}
+        {{ $key }}: {{ toYaml $val | indent 8 | trimPrefix (repeat 8 " ") }}
+        {{- end }}
+        {{- end }}
+        {{- with $templateData.pod.labels }}
+        {{- range $key, $val := . }}
+        {{ $key }}: {{ toYaml $val | indent 8 | trimPrefix (repeat 8 " ") }}
+        {{- end }}
+        {{- end }}
+      generateName: {{ toYaml .Job.generateName }}
+    spec:
+      {{- if .Job.host }}
+      nodeName: {{ toYaml .Job.host }}
+      {{- end }}
+      restartPolicy: Never
+      {{- with $templateData.pod.serviceAccountName }}
+      serviceAccountName: {{ . }}
+      {{- end }}
```

Comment:
@colearendt I think we will need to deprecate `pod.serviceAccountName` in favor of the Connect setting `config.Launcher.KubernetesDefaultServiceAccount` since the default launcher template now includes support for service accounts.

As it stands today, setting both `pod.serviceAccountName` and `config.Launcher.KubernetesDefaultServiceAccount` would result in 2 separate `serviceAccountName` files in the pod spec which is probably invalid.

Reply:
Nevermind, I misunderstood how these values were getting mapped to the content pod.

Instead, we should re-purpose `launcher.templatesValues.pod.serviceAccountName` to set `config.Launcher.KubernetesDefaultServiceAccount`

Reply:
That's a great call! I think the template here kinda needs to support both b/c Workbench (for instance) doesn't have a way to pass per-job service account stuff. It's also nice to have as a "fallback." Does that seem like the right call for precedence? Will Connect _always_ be sending a service account? Or does it leave that alone / empty in some cases / if not configured?

I figure we can probably do something like this (presuming I get the `default` precedence right... 😬 ):

```
{{- if or $templateData.pod.serviceAccountName .Job.serviceAccountName }}
serviceAccountName: {{ .Job.serviceAccountName | default $templateData.pod.serviceAccountName }}
{{- end }}
```

Reply:
sorry for the delay here - getting caught up after the holidays - 

Connect will not always be sending a service account. If one isn't explicitly configured, Connect won't send a service account through.

Reply:
@colearendt 

> Will Connect always be sending a service account? Or does it leave that alone / empty in some cases / if not configured?

No, Connect will not always set the default service account. If none is set, then we don't set ServiceAccountName on the JobConfig meaning Kubernetes will assign `default` to the Pod.

Is `.Job.serviceAccountName` the same value as `config.Launcher.KubernetesDefaultServiceAccount` in this context? If so, then I think what you have here is correct.

Then I think the only addition we will need in this PR is to set a default value in the `values.yaml` to be `launcher.templateValues.pod.serviceAccountName: false`? Does that sound right?

Reply:
Sorry, I got confused about the template's context here. `.Job` in this context is the `JobConfig` that Connect sends to the launcher. So yes, I think the precedence here is correct.

Reply:
I think I have taught you a bad convention there. I have since learned that `""` is a better default than `false` 😄 

The thing we should probably handle in the templates though (hence my `if` recommended above) is the case where _both_ are defined. We don't want to end up with two `serviceAccountName`s on the job spec 😬

Reply:
@colearendt  How's this? https://github.com/rstudio/helm/pull/325/commits/c526c6a134f0a0143fa420bd90842c723bd36ca9

I think this should be ready to merge anytime now

Reply:
That looks good to me! I would just check to be sure that precedence and defaults work as expected. 😄

Reply:
@zackverham  @colearendt Note that doing it this way may lead to some future confusion. 

For example, if I were to set the following:

```
config:
  Launcher:
    KubernetesDefaultServiceAccount: "my-default"
launcher:
  templateValues:
    pod:
      serviceAccountName: "different-service-account"
```

Then my template data contains `different-service-account`: 

```
data:
  rstudio-library-templates-data.tpl: |
    {{- define "rstudio-library.templates.data" -}}
    {"job":{"annotations":{},"labels":{}},"pod":{"affinity":{},"annotations":{},"containerSecurityContext":{},"defaultSecurityContext":{},"extraContainers":[],"imagePullPolicy":"","imagePullSecrets":[],"initContainers":[{"image":"ghcr.io/rstudio/rstudio-connect-content-init:bionic-2022.12.0","imagePullPolicy":"IfNotPresent","name":"init","securityContext":{},"volumeMounts":[{"mountPath":"/mnt/rstudio-connect-runtime/","name":"rsc-volume"}]}],"labels":{},"securityContext":{},"serviceAccountName":"different-service-account","tolerations":[],"volumeMounts":[{"mountPath":"/opt/rstudio-connect","name":"rsc-volume"}],"volumes":[{"emptydir":{},"name":"rsc-volume"}]},"service":{"annotations":{},"labels":{},"type":"ClusterIP"}}
    {{- end }}
```

and my `rstudio-connect.gcfg` contains:

```
  [Launcher]
  KubernetesDefaultServiceAccount = "my-default"
```

Meaning that `.Job.ServiceAccount` will be `my-default`. Everything will work as expected but it may confuse someone later so I'm just tracking this for future us.

Reply:
Yeah, that makes sense. It is an unfortunate happenstance 😛 I'm thinking about possible ways to mitigate... would it be possible to embed a comment that would make it into the output YAML? `# default may be overridden by the application` or something?